protocol/binary/stream_writer.go

```diff
@@ -0,0 +1,251 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package binary
+
+import (
+	"io"
+	"math"
+	"sync"
+
+	"go.uber.org/thriftrw/internal/iface"
+	"go.uber.org/thriftrw/protocol/stream"
+)
+
+var streamWriterPool = sync.Pool{
+	New: func() interface{} {
+		writer := &StreamWriter{}
+		return writer
+	}}
+
+// StreamWriter implements basic logic for writing the Thrift Binary Protocol
+// to an io.Writer.
+type StreamWriter struct {
+	// Private implementation to disallow custom implementations of
+	// the Writer interface
+	iface.Impl
+
+	writer io.Writer
+
+	// This buffer is re-used every time we need a slice of up to 8 bytes.
+	buffer [8]byte
+}
+
+// BorrowStreamWriter fetches a StreamWriter from the system that will write
+// its output to the given io.Writer.
+//
+// This StreamWriter must be returned back using ReturnStreamWriter.
+func BorrowStreamWriter(w io.Writer) *StreamWriter {
+	streamWriter := streamWriterPool.Get().(*StreamWriter)
+	streamWriter.writer = w
+	return streamWriter
+}
+
+// ReturnStreamWriter returns a previously borrowed StreamWriter back to the
+// system.
+func ReturnStreamWriter(sw *StreamWriter) {
+	sw.writer = nil
+	streamWriterPool.Put(sw)
+}
+
+func (bw *StreamWriter) write(bs []byte) error {
+	_, err := bw.writer.Write(bs)
+	return err
+}
+
+func (bw *StreamWriter) writeByte(b byte) error {
+	bs := bw.buffer[0:1]
+	bs[0] = b
+	return bw.write(bs)
+}
+
+func (bw *StreamWriter) writeInt16(n int16) error {
+	bs := bw.buffer[0:2]
+	bigEndian.PutUint16(bs, uint16(n))
+	return bw.write(bs)
+}
+
+func (bw *StreamWriter) writeInt32(n int32) error {
+	bs := bw.buffer[0:4]
+	bigEndian.PutUint32(bs, uint32(n))
+	return bw.write(bs)
+}
+
+func (bw *StreamWriter) writeInt64(n int64) error {
+	bs := bw.buffer[0:8]
+	bigEndian.PutUint64(bs, uint64(n))
+	return bw.write(bs)
+}
+
+func (bw *StreamWriter) writeString(s string) error {
+	if err := bw.writeInt32(int32(len(s))); err != nil {
+		return err
+	}
+
+	_, err := io.WriteString(bw.writer, s)
+	return err
+}
+
+// WriteBool encodes a boolean
+func (bw *StreamWriter) WriteBool(b bool) error {
+	if b {
+		return bw.writeByte(1)
+	}
+	return bw.writeByte(0)
+}
+
+// WriteInt8 encodes an int8
+func (bw *StreamWriter) WriteInt8(i int8) error {
+	return bw.writeByte(byte(i))
+}
+
+// WriteInt16 encodes an int16
+func (bw *StreamWriter) WriteInt16(i int16) error {
+	return bw.writeInt16(i)
+}
```

Comment:
Do we need these exported wrappers?  Could we export the underlying write instead, the `Writer` would then call the exported methods?